code/obj/flock/structure/sapper.dm

```diff
@@ -0,0 +1,147 @@
+#define CHARGING_BITS "Speed up Flockbits"
+#define CHARGING_DRONES "Charge incapacitors"
+#define CHARGING_STRUCTURES "Enhance structures"
+
+/obj/flock_structure/sapper
+	icon_state = "sapper-off" // current sprites are placeholders
+	name = "sparking machine"
+	desc = "A strange structure supporting four coils. You can feel the electricity coming off of it"
+	flock_desc = "A multipurpose support structure that saps power from a local APC. It can be set to enchance Flockbits, Flockdrones, or structures."
+	flock_id = "Sapper"
+	health = 30
+	health_max = 30
+	repair_per_resource = 1.5
+	resourcecost = 100
+
+	passthrough = TRUE
+
+	var/mode = CHARGING_BITS
+	var/mode_cooldowns = list(CHARGING_BITS = 20 SECONDS, CHARGING_DRONES = 30 SECONDS, CHARGING_STRUCTURES = 30 SECONDS)
+
+	var/obj/machinery/power/apc/linked_apc = null
+
+	New(atom/location, datum/flock/F = null)
+		..()
+		src.linked_apc = src.try_link_apc()
+		if (!src.linked_apc)
+			src.icon_state = "sapper-off"
+
+		src.info_tag.set_info_tag("Mode: [src.mode]")
+
+		ON_COOLDOWN(src, CHARGING_BITS, src.mode_cooldowns[CHARGING_BITS])
+		ON_COOLDOWN(src, CHARGING_DRONES, src.mode_cooldowns[CHARGING_DRONES])
+		ON_COOLDOWN(src, CHARGING_STRUCTURES, src.mode_cooldowns[CHARGING_STRUCTURES])
+
+	building_specific_info()
+		return {"<span class='bold'>Mode:</span> [src.mode].
+				<br><span class='bold'>Linked power supply charge:</span> [src.linked_apc?.cell ? "[round(src.linked_apc.cell.charge / src.linked_apc.cell.maxcharge * 100)]%": "Not linked"]."}
+
+	process(mult)
+		if (!src.linked_apc)
+			src.try_link_apc()
+			if (!src.linked_apc)
+				src.icon_state = "sapper-off"
+				return
+		if (!linked_apc.cell || linked_apc.cell.charge <= 0 || linked_apc.status & BROKEN)
+			src.icon_state = "sapper-off"
+			return
+
+		src.icon_state = "sapper-on"
+
+		if (GET_COOLDOWN(src, src.mode))
+			return
+		if (linked_apc.cell.charge < linked_apc.cell.maxcharge * 0.1)
+			return
+
+		var/list/targets = list()
+
+		switch(src.mode)
+			if (CHARGING_BITS)
+				var/list/nearby_bits = list()
+				var/area/src_area = get_area(src)
+				var/perform_lag_check = length(src.flock.units[/mob/living/critter/flock/bit]) > 10
+				for (var/mob/living/critter/flock/bit/flockbit as anything in src.flock.units[/mob/living/critter/flock/bit])
+					if (perform_lag_check)
+						LAGCHECK(LAG_MED)
+					if (GET_DIST(src, flockbit) < 7 || get_area(flockbit) == src_area)
```

Comment:
Seems strange that it applies to a whole area considering that areas are weirdly defined in normal gameplay. "Be within 7 tiles to get charged" makes a lot more sense than "be in the same vague area to get charged".
Is the idea to be a 'flock APC' kinda?

Reply:
Yeah that was the original idea, original implementation powered Flockdrones and structures in the same area too but it was changed around. Went with removing the area check